flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedDistinctRowsOver.scala

```diff
@@ -0,0 +1,230 @@
+      new MapStateDescriptor[Long, JList[Row]]("windowBufferMapState",
+        BasicTypeInfo.LONG_TYPE_INFO.asInstanceOf[TypeInformation[Long]], rowListTypeInfo)
+    rowMapState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+    val aggregationStateDescriptor: ValueStateDescriptor[Row] =
+      new ValueStateDescriptor[Row]("aggregationState", aggregatesTypeInfo)
+    accumulatorState = getRuntimeContext.getState(aggregationStateDescriptor)
+
+    val processedCountDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("processedCountState", classOf[Long])
+    counterState = getRuntimeContext.getState(processedCountDescriptor)
+
+    val smallestTimestampDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("smallestTSState", classOf[Long])
+    smallestTsState = getRuntimeContext.getState(smallestTimestampDescriptor)
+    distinctValueStateList = new Array(aggregates.size)
+    for(i <- 0 until aggregates.size){
+      if(distinctAggsFlag(i)){
+        val distinctValDescriptor =  new MapStateDescriptor[Any, Long](
+                                             "distinctValuesBufferMapState" + i,
+                                              classOf[Any],
+                                              classOf[Long])
+        distinctValueStateList(i)=getRuntimeContext.getMapState(distinctValDescriptor)
+      }
+    }
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+
+    val currentTime = ctx.timerService.currentProcessingTime
+    var i = 0
+
+    // initialize state for the processed element
+    var accumulators = accumulatorState.value
+    if (accumulators == null) {
+      accumulators = new Row(aggregates.length)
+      while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      }
+    }
+
+    // get smallest timestamp
+    var smallestTs = smallestTsState.value
+    if (smallestTs == 0L) {
+      smallestTs = currentTime
+      smallestTsState.update(smallestTs)
+    }
+    // get previous counter value
+    var counter = counterState.value
+
+    if (counter == precedingOffset) {
+      val retractList = rowMapState.get(smallestTs)
+
+      // get oldest element beyond buffer size
+      // and if oldest element exist, retract value
+      var removeCounter :Integer = 0
+      var distinctCounter : Integer = 0
+      var retractVal : Object = null
+      i = 0
+      while (i < aggregates.length) {
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        retractVal = retractList.get(0).getField(aggFields(i)(0))
+        if(distinctAggsFlag(i)){
+          var distinctValCounter: Long = distinctValueStateList(i).get(retractVal)
+          // if the value to be retract is the last one added
+          // the remove it and retract the value
+          if(distinctValCounter == 1L){
```

Comment:
if we are retracting, it cannot be null, right? it is a value that was added to the list before at least once. I get your point thou